### PR TITLE
ZEPPELIN-299 Support clearing output for paragraph

### PR DESCRIPTION
**Pull request by r-kamath**
add an option to clear output/result for paragraph on UI and necessary backend changes

clear-output-notebook
--
clear-output-paragraph

### CODE REVIEW

File: zeppelin-web/src/app/notebook/notebook.controller.js [hunk 1]
```diff
@@ -104,6 +104,15 @@ angular.module('zeppelinWebApp').controller('NotebookCtrl', function($scope, $ro
     $scope.isNoteDirty = null;
   };
 
+  $scope.clearAllParagraphOutput = function() {
+    var result = confirm('Do you want to clear all output?');
+    if (result) {
+      _.forEach($scope.note.paragraphs, function(n, key) {
+        angular.element('#' + n.id + '_paragraphColumn_main').scope().clearParagraphOutput(n.id);
```
Comment (corneadoug) [inline zeppelin-web/src/app/notebook/notebook.controller.js]: clearParagraphOutput doesn't take any parameter

**Comment by corneadoug** [comment]
How would this affect running paragraphs?
Do we need a warning?

**Comment by r-kamath** [comment]
@corneadoug cleaned up the unused param in a311fce
> How would this affect running paragraphs?

clear output button is disabled when a paragraph is running

**Comment by corneadoug** [comment]
The button at notebook level is disabled, not those at paragraph level.

**Comment by r-kamath** [comment]
Not sure what's the right way. Disable the link or show another confirm box ?

**Comment by corneadoug** [comment]
The problem is to handle the running job in the backend.
If a job is running or queued, they need to be stopped otherwise they might print result after we cleared output.

The easiest way then, would be to hide that action in both paragraph and Notebook if something is running. (Hide in Notebook if anything is running, Hide in paragraph if that paragraph is running or queued)

**Comment by r-kamath** [comment]
@corneadoug :+1: for hiding the option.
> (Hide in Notebook if anything is running, Hide in paragraph if that paragraph is running or queued)

I think it should be "hide" if anything is running or queued, if we hide only in paragraph when the that paragraph is running, the user will still be able to call clear all from notebook

**Comment by corneadoug** [comment]
Yes, that's what I said:
- Hide the 'clear all notebook button' if any paragraph is running or queued
- Hide the 'clear this paragraph button' if that paragraph is running or queued

This would allow us to clear at paragraph level on other non-running paragraph

**Comment by prabhjyotsingh** [comment]
@corneadoug  I believe this is a default behaviour that we do on terminal (command line) as well. After firing a huge command which will result in a lots of log; run the particular command and press Ctrl-L to clear the screen. 
Or even in case of firebug console as well, console logs keep flowing, and we do clear log, just to see what appeared next.
Hence, personally I don't feel a need of either disabling or removing or giving a confirmation box.

**Comment by corneadoug** [comment]
@prabhjyotsingh the goal of that issue was to allow people to clean the output of their query from the notebook, especially for demo purpose.
Also, zeppelin get the result once, it is not a live flowing message stack.

**Comment by prabhjyotsingh** [comment]
@corneadoug  idea of clear button is to clear the output, whenever the user wants. And not necessarily for demo purposes.
Now its up to the user when (s)he wants.

Example: Re-run of notebook with code change few of the paragraphs.

Usage of too many alert/confirm boxes are bad UX.

**Comment by Leemoonsoo** [comment]
Thanks @prabhjyotsingh for nice new feature.
To me, even clearing output in RUNNING or PENDING status is useful.
Looks good to me.

**Comment by rconline** [comment]
Looks good, useful feature.

**Comment by vinayshukla** [comment]
Is this feature included in Zeppelin 0.5.5 RC1?

**Comment by felixcheung** [comment]
@vinayshukla no since this PR is still open...

**Comment by Leemoonsoo** [comment]
Merge if there're no more discussions

**Pull request closed without merging**